### PR TITLE
mTLS Kong Admin API authentication support #1958

### LINKED ISSUE #1958: mTLS Kong Admin API authentication support
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem Statement

This is a new feature request to support Kong mTLS authentication capabilities.

Kong API administration API can be secured via mTLS authentication by setting the following NGINX directives,

```
    nginx_admin_proxy_ssl_verify: "on"
    nginx_admin_ssl_client_certificate: /etc/secrets/tls.crt
    nginx_admin_ssl_verify_client: "on"
    nginx_admin_ssl_verify_depth: 3
    nginx_admin_ssl_trusted_certificate: /etc/secrets/ca/ca-bundle.crt
```

Thus, the ingress controller should provide a way to specify client certificates.

### Proposed Solution

Add new arguments to specify the client certificate and key and setup the http/client package to use client certificates.

### Additional information

_No response_

### Acceptance Criteria

Once mTLS activated the ingress controller should authenticate to Kong administrative API with the specified client certificate and key.

### PR DESCRIPTION
**Pull request by ludovic-pourrat**
Feature request proposal related to mTLS Kong Admin API authentication support #1958

**What this PR does / why we need it**:

To support Kong administrative API mTLS authentication.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: 

fixes #1958

**Special notes for your reviewer**:

Documentation is missing and testing advise is required.

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [ ] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

**Review by shaneutt — DISMISSED** [review]
First off thank you for taking this issue on and contributing we appreciate it :vulcan_salute:

I did a first pass review but I think things are likely to change significantly after tests are added so let me just address your comments from the description:

> documentation is missing

Documentation for this will need to be done as a separate PR to https://github.com/kong/docs.konghq.com, and should include updates to pages like:

https://docs.konghq.com/kubernetes-ingress-controller/2.0.x/concepts/security/#authentication-on-kongs-admin-api

And other places where `mTLS` for the Admin API is mentioned.

This can be done _after_ this PR.

> and testing advise is required

Other than unit tests, we have `test/integration` and `test/e2e` the former being integration tests for standard features in a microcosm where standard controller manager options are applied and the latter being a complete cluster stand up, testing performed, and cluster torn down where custom controller manager options can be provided. Given that context the `test/e2e` tests will be a good place to start.

Some additional context:

Testing is highly integrated with our [Kong Kubernetes Testing Framework (KTF)](https://github.com/kong/kubernetes/kubernetes-testing-framework) which does cluster and component provisioning for us and provisioning is generally structured like this:

- a `kind` cluster is deployed using KTF
- Kong Gateway is deployed using the `kong` KTF cluster addon flagged with `WithControllerDisabled()` to deploy without the controller
- The controller is deployed separately and locally and configured to use the Admin API of the Gateway deployed in the cluster, flagged however you like

Some examples of this can be seen in `test/e2e` as well as in the [integration tests in the KTF Repo](https://github.com/kong/kubernetes-testing-framework/tree/main/test/integration) to get you started.

In addition to the above, your specific test will likely have some steps like this:

- generate a CA
- generate server and client certificates signed by the CA
- configure the `kong` KTF addon for `mTLS`
- configure the controller manager for `mTLS`
- some basic tests that trigger data-plane configurations over the `mTLS` connection (e.g. deploy `Ingress` and ensure routes get configured properly on the data-plane, there are several examples)

Also there's some additional notes/caveats related to this you should know:

- currently the testing environments from KTF [only support Linux](https://github.com/Kong/kubernetes-testing-framework/), this is a consequence of current maintainers all being Linux developers. If you use Mac/Windows and would be interested in adding support let us know and we can help support that and give you more context.
- it will likely be necessary to add options to KTF to enable `mTLS` on the `kong` addon (see `pkg/clusters/addons/kong`), this is something I can help you with and provide more context, let me know your thoughts.
- it may be feasible to use [Kubernetes native certificate signing requests](https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/) and rely on the Kubernetes API's CA in your test for convenience, though we don't have precedent for this in the tests currently so some experimentation will be required.

Please feel free to ping me or @Kong/team-k8s directly in this PR for help along the way if the existing tests aren't example enough or you otherwise get stuck.

**Comment by rainest** [comment]
After some internal discussion, we think a simpler test is viable here, since the bulk of the client certificate handling is implemented upstream (in Golang `http.Client` and in NGINX) and the controller is only binding certificates to the `http.Client`.

If you haven't already done much of the work to implement an integration test, can you instead write a unit test that passes an `HTTPClientOps` with client certificate fields set to `MakeHTTPClient()` and then confirm that the resulting `http.Client` does indeed have those certificates set within its `Transport`?

File: internal/manager/config.go [hunk 1]
```diff
@@ -110,6 +110,9 @@ func (c *Config) FlagSet() *pflag.FlagSet {
 	flagSet.BoolVar(&c.EnableReverseSync, "enable-reverse-sync", false, `Send configuration to Kong even if the configuration checksum has not changed since previous update.`)
 	flagSet.DurationVar(&c.SyncPeriod, "sync-period", time.Hour*48, `Relist and confirm cloud resources this often`) // 48 hours derived from controller-runtime defaults
 
+	flagSet.StringVar(&c.KongAdminAPIConfig.TLSClientKey, "kong-admin-tls-client-key", "", "mTLS client key for authentication.")
+	flagSet.StringVar(&c.KongAdminAPIConfig.TLSClientCert, "kong-admin-tls-client-cert", "", "mTLS client certificate for authentication.")
+
```
[the remark this answers is not in the document]
Comment (ludovic-pourrat) [reply internal/manager/config.go:115]: Make sense I will update accordingly

**Comment by ludovic-pourrat** [comment]
@shaneutt and @rainest working on it this week-end to integrate your feedbacks, a new PR will follow.

Thanks !

**Pull request closed without merging**